dmipy/utils/invariants.py

```diff
@@ -0,0 +1,576 @@
+    -------
+    invariants : array
+        array containing the desired invariants
+    """
+    if not(sh_base in ['mrtrix', 'dipy']):
+        raise ValueError(
+            'sh_base must be either "mrtrix" or "dipy"')
+
+
+    l_max = np.max([np.max(l) for l in l_list])
+    if l_max > 8:
+        raise ValueError(
+            'maximum degree in l_list is greater than 8')
+
+    len_max = np.max([len(l) for l in l_list])
+    if len_max > 5:
+        raise ValueError(
+            'maximum power in l_list is greater than 5')
+
+    odd_number = np.sum([np.sum(np.array(l)%2) for l in l_list])
+    if odd_number > 0:
+        raise ValueError(
+            'all of the degrees in l_list must be even')
+
+    N = sh_coef.shape[-1]
+    sh_order = int((-3 + np.sqrt(1+8*N))/2.0)
+
+    if sh_order < l_max:
+        raise ValueError(
+            'SH order of the coefficients is lower than the degree of the required invariants')
+
+    RG = load_npz(join(DATA_PATH, 'real_gaunt_coefficients.npz')).toarray()[0,:]
+    coef = convert_sh(sh_coef, in_base=sh_base)
+
+    inv_number = len(l_list)
+
+    if len(coef.shape) > 1:
+        shape = list(coef.shape[:-1])+[inv_number]
+    else:
+        shape = inv_number
+
+    invariants = np.zeros(shape)
+
+    s4p = np.sqrt(4*np.pi)
+    is4p = 1.0/np.sqrt(4*np.pi)
+    MTV = []
+    for i, l_i in enumerate(l_list):
+        len_l = len(l_i)
+        if len_l==1:
+            if l_i[0] == 0:
+                mtv = 1
+                invariants[..., i] = coef[...,0]*s4p
+
+
+        if len_l==2:
+            if l_i[0] == l_i[1]:
+                ll = l_i[0]
+                mtv = (2*ll+1)/(4*np.pi)
+                start = (ll-1)*(ll)//2
+                stop = (ll+1)*(ll+2)//2
+                invariants[...,i] = np.sum(coef[..., start:stop]**2, -1)
+
+        if len_l==3:
+            l1 = l_i[0]
+            l2 = l_i[1]
+            l3 = l_i[2]
+            mtv = 0
+
+            for m1 in range(-l1,l1+1):
+                for m2 in range(-l2,l2+1):
+                    for m3 in range(-l3,l3+1):
+                        rg = real_gaunt(l1,l2,l3,m1,m2,m3,RG)
+                        mtv += gaunt_real_sph_harm(m1,l1,0,0) * gaunt_real_sph_harm(m2,l2,0,0) * gaunt_real_sph_harm(m3,l3,0,0)* rg
+                        invariants[...,i] += coef[...,_get_index(l1,m1)]*coef[...,_get_index(l2,m2)]*coef[...,_get_index(l3,m3)]*rg
+
+        if len_l==4:
+            l1 = l_i[0]
+            l2 = l_i[1]
+            l3 = l_i[2]
+            l4 = l_i[3]
+            mtv = 0
+            rsh=[]
+            for m1 in range(-l1,l1+1):
+                for m2 in range(-l2,l2+1):
+                    for m3 in range(-l3,l3+1):
+                        for m4 in range(-l4,l4+1):
+                            rg = real_gaunt4(l1,l2,l3,l4,m1,m2,m3,m4,RG)
+                            mtv += gaunt_real_sph_harm(m1,l1,0,0) * gaunt_real_sph_harm(m2,l2,0,0)*gaunt_real_sph_harm(m3,l3,0,0)*gaunt_real_sph_harm(m4,l4,0,0)*rg
```

Comment:
I think in this and the following cases you can replace this long manual line with the gaunt multiplication with something like `[gaunt_real_sph_harm(m, l, 0, 0) for m, l in zip([m1,m2],[l1,l2])]` and then do the multication with `rg` on a new line.

Reply:
Yeah, maybe I can use itertools.product or something like that. I need to try it and test if it works.

Reply:
Actually, am I right that you are doing many double calculations of these `gaunt_real_sph_harm(m, l, 0, 0)` guys for the same `l,m` in all your for loops?

Would it not be much more efficient to precompute the unique gaunt `m,l` combinations into a `gaunt_dict` before the main for-loop, and then just call them from the `dict`in these bigger for-loops?

Reply:
Alright already much better! 

Now we can go for the gold and actually merge the 3,4 and 5 cases in a single statement.

Try something like:
```
If len_l > 2:
    m_list = itertools.product(*[range(-l, l+1) for l in l_i])
    for m_j in m_list:
        # choose based in len(l_i)
        rg = real_gauntn(*l_i, *m_j, RG)
        tmp_mtv = [g_sh[_get_index(l, m)] for l, m in zip(l_i, m_j)]
        mtv += operator.mul(tmp_mtv) * rg     
        tmp_invariants= [coef[...,_get_index(l, m)] for l, m in zip(l_i, m_j)]
        invariants[..., i] += operator.mul(tmp_invariants) * rg
```